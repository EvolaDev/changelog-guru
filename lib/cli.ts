import commander, { Command } from 'commander';
import fs from 'fs';
import path from 'path';
import packageJson from '../package.json';
import Process from './process';

export default class CLI {
    public static CONFIG_FILE_NAME: string = '.changelog';
    public static ENV_TOKEN_NAME: string = 'CHANGELOG_GURU_TOKEN';

    public static help(): void {
        commander.help();
    }

    public static showHelp(msg: string): void {
        Process.error(msg, false);
        CLI.help();
        Process.exit();
    }

    public static parse(): string[] {
        const command: Command = commander.parse(Process.getArguments());
        const getPath = (pathSegment: string): string => path.resolve(Process.CWD, pathSegment);
        const configPath: string = getPath(command.config || path.join(__dirname, `../${CLI.CONFIG_FILE_NAME}`));
        const packagePath: string = getPath(command.package || 'package.json');
        const token: string = command.token || Process.getVariable(CLI.ENV_TOKEN_NAME);

        if (!fs.existsSync(configPath)) CLI.showHelp('<config> options is not an existing filename');
        if (!fs.existsSync(packagePath)) CLI.showHelp('<package> package.json not found');
        if (!token) CLI.showHelp('<token> options must be provided');

        return [configPath, packagePath, token];
    }
}

commander.version(packageJson.version, '-v, --version');
commander.description(packageJson.description);
commander.option('-c, --config <config>', `config file in JSON format (${CLI.CONFIG_FILE_NAME}).`);
commander.option('-p, --package <package>', `path to package.json.`);
commander.option('-t, --token <token>', `your GitHub access token (${CLI.ENV_TOKEN_NAME} by default).`);
