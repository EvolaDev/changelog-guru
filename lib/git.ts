import rest, { ReposListCommitsResponseItem, ReposListReleasesParams } from '@octokit/rest';
import Process from './process';

export default class Git {
    public static COMMITS_PAGE_SIZE: number = 100;
    private kit: rest;
    private repository: ReposListReleasesParams;
    private sha: string;

    public constructor(repo: string, owner: string, token: string, sha: string) {
        this.repository = { repo, owner } as ReposListReleasesParams;
        this.sha = sha;
        this.kit = new rest({ auth: `token ${token}` });

        Process.info('Repository', this.repository.repo);
        Process.info('Owner', this.repository.owner);
        Process.info('SHA', this.sha);
    }

    public async getSince(): Promise<string> {
        const { data: { length } } = await this.kit.repos.listReleases({ ...this.repository });
        let since: string = (new Date(0)).toISOString();

        if (length) {
            const { data: release } = await this.kit.repos.getLatestRelease({ ...this.repository });

            since = release.created_at;
        }

        Process.info('Get last commits since', since);

        return since;
    }

    public async getCommits(since: string, page: number): Promise<ReposListCommitsResponseItem[]> {
        const { data: commits } = await this.kit.repos.listCommits({
            since,
            ...this.repository,
            sha: this.sha,
            ...{ page, per_page: Git.COMMITS_PAGE_SIZE },
        });

        return commits;
    }
}
