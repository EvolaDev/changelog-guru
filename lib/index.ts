import Reader from './src/reader';

const reader = new Reader();

reader.init();
reader.parse().then(() => {
    process.exit(0);
});
