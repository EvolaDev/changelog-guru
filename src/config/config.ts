import { TaskTree } from 'tasktree-cli';
import { ChangeLevel, ExclusionType, ServiceProvider } from './typings/enums';
import { ConfigOptions, PluginOption } from './typings/types';
import { Provider } from '../providers/provider';
import { GitHubProvider } from '../providers/github/provider';

export class Config {
    public readonly filePath: string;
    public readonly provider: ServiceProvider;

    private types: Map<string, ChangeLevel>;
    private plugins: Map<string, PluginOption>;
    private exclusions: Map<ExclusionType, string[]>;

    public constructor(options: ConfigOptions) {
        this.provider = options.provider;
        this.filePath = options.filePath;
        this.types = options.types;
        this.plugins = options.plugins;
        this.exclusions = options.exclusions;
    }

    public async getProvider(repository: string): Promise<Provider> {
        const tasks = TaskTree.tree();
        let provider: Provider | undefined;

        switch (this.provider) {
            case ServiceProvider.GitLab:
                tasks.fail(`${ServiceProvider.GitLab} - not supported yet`);
                break;
            case ServiceProvider.GitHub:
                provider = new GitHubProvider(repository);
                break;
            default:
                tasks.fail(`Service provider not specified`);
                break;
        }

        return provider as Provider;
    }

    public getPlugin(name: string): PluginOption | undefined {
        return this.plugins.get(name);
    }

    public getPlugins(): [string, PluginOption][] {
        return [...this.plugins.entries()];
    }

    public getTypes(): [string, ChangeLevel][] {
        return [...this.types.entries()];
    }

    public getExclusions(): [ExclusionType, string[]][] {
        return [...this.exclusions.entries()];
    }
}
