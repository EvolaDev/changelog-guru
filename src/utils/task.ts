import chalk from 'chalk';
import logSymbols from 'log-symbols';
import figures from 'figures';
import elegantSpinner from 'elegant-spinner';
import indentString from 'indent-string';
import Process from './process';

const spinner = elegantSpinner();

export enum Status {
    Pending,
    Completed,
    Failed,
    Skipped
}

export default class Task {
    public static LINE_SEPARATOR: string = '\n';
    public static LINE_INDENT: { indent: string } = { indent: '  ' };
    public static SYMBOL_GROUP: string = chalk.yellow(figures.pointer);
    public static SYMBOL_SUCCESS: string = logSymbols.success;
    public static SYMBOL_SKIPPED: string = chalk.yellow(figures.arrowDown);
    public static SYMBOL_ERROR: string = logSymbols.error;
    public static SYMBOL_WARNING: string = logSymbols.warning;
    public static SYMBOL_INFO: string = logSymbols.info;
    public static SYMBOL_PREFIX: string = `${chalk.dim(figures.arrowRight)} `;

    public static SYMBOL_EMPTY = '';

    private text: string;
    private status: Status;
    private subtasks: Task[] = [];
    private logs: string[] = [];
    private warnings: string[] = [];

    public constructor(text: string, status: Status = Status.Pending) {
        this.text = text;
        this.status = status;
    }

    private static renderList(list: string[], symbol: string, level: number): string[] {
        return list.map((text: string): string => indentString(`${symbol} ${text}`, level + 1, Task.LINE_INDENT));
    }

    public add(text: string, status: Status = Status.Pending): Task {
        const task: Task = new Task(text, status);

        this.subtasks.push(task);

        return task;
    }

    public complete(text?: string): void {
        if (!this.subtasks.filter((task): boolean => task.isPending()).length) {
            this.update(Status.Completed, text);
        } else {
            this.fail('Subtasks is not complete.');
        }
    }

    public skip(text?: string): void {
        this.update(Status.Skipped, text);
    }

    public fail(error?: string): void {
        this.update(Status.Failed, this.getError(error));
        Process.getInstance().end(false);
    }

    public log(text: string): void {
        if (text) this.logs.push(text);
    }

    public warn(text: string): void {
        if (text) this.warnings.push(text);
    }

    public isPending(): boolean {
        return this.status === Status.Pending;
    }

    public getActive(): Task {
        const { subtasks } = this;
        const subtask: Task | undefined = subtasks[subtasks.length - 1];
        let task: Task = this;

        if (subtask && subtask.isPending()) {
            task = subtask.getActive();
        }

        return task;
    }

    public render(level: number = 0): string {
        const skipped = this.status === Status.Skipped ? ` ${chalk.dim('[skipped]')}` : Task.SYMBOL_EMPTY;
        const prefix = level ? Task.SYMBOL_PREFIX : Task.SYMBOL_EMPTY;
        const text = [
            indentString(`${prefix}${this.getSymbol()} ${this.text}${skipped}`, level, Task.LINE_INDENT),
            ...Task.renderList(this.warnings, Task.SYMBOL_WARNING, level),
            ...Task.renderList(this.logs, Task.SYMBOL_INFO, level),
            ...this.subtasks.map((task: Task): string => task.render(level + 1))
        ].join(Task.LINE_SEPARATOR);

        return level ? chalk.dim(text) : text;
    }

    private getSymbol(): string {
        let symbol: string;

        switch (this.status) {
            case Status.Skipped:
                symbol = Task.SYMBOL_SKIPPED;
                break;
            case Status.Completed:
                symbol = Task.SYMBOL_SUCCESS;
                break;
            case Status.Failed:
                symbol = Task.SYMBOL_ERROR;
                break;
            default:
                symbol = this.subtasks.length ? Task.SYMBOL_GROUP : chalk.yellow(spinner());
                break;
        }

        return symbol;
    }

    private getError(error?: string): string {
        return error ? `${this.text}: ${chalk.redBright(error)}` : this.text;
    }

    private update(status: Status, text?: string): void {
        if (this.isPending()) {
            if (text) this.text = text;

            this.status = status;
        } else {
            this.text = this.getError(
                `Trying to change the status of a completed task (${chalk.bold(this.status.toString())})`
            );
            this.status = Status.Failed;
        }
    }
}
