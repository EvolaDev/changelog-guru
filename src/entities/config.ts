import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import chalk from 'chalk';
import Process from '../utils/process';
import { ProviderName } from '../providers/provider';
import { ReadonlyArray, Option, OptionValue } from '../utils/types';

const $process = Process.getInstance();
const fileName = '.changelog.yaml';
const load = (p: string): ConfigOptions => yaml.safeLoad(fs.readFileSync(p, 'utf8')) || {};
const defaultConfig: ConfigOptions = load(path.join(__dirname, '../../', fileName));

export interface ConfigOptions extends Option {
    config?: string;
    types?: string[];
    plugins?: string[];
    provider?: ProviderName;
    [key: string]: Option | OptionValue;
}

export default class Config {
    public readonly plugins: ReadonlyArray<string> = defaultConfig.plugins || [];
    public readonly types: ReadonlyArray<string>;
    public readonly options: Option;
    public readonly provider: ProviderName;

    public constructor(options?: ConfigOptions) {
        $process.addTask('Config initializing');

        let config: ConfigOptions = Object.assign({}, defaultConfig, options || {});
        let { config: filePath } = config;

        if (typeof filePath === 'string') {
            filePath = path.resolve(process.cwd(), filePath);

            if (typeof filePath === 'string' && filePath.length && fs.existsSync(filePath)) {
                config = Object.assign({}, config, load(filePath));

                $process.addSubTask(`Used config file from: ${filePath}`, true);
            } else {
                $process.addSubTask(`File ${chalk.bold(fileName)} is not exists`, true);
            }
        } else {
            $process.addSubTask(`File path to ${chalk.bold(fileName)} is not specified`, true);
            $process.addSubTask(`Used default config`);
        }

        this.provider = config.provider || ProviderName.GitHub;
        this.types = config.types || [];
        this.plugins = this.plugins.concat(config.plugins || []);
        this.options = new Proxy(config as Option, {
            get(target, name, receiver): Option | undefined {
                return Reflect.get(target, name, receiver);
            }
        });

        $process.completeTask();
    }
}
