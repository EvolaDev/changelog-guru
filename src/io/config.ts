import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml'
import { ReadonlyArray } from '../utils/types';
import { ProviderName } from './providers/provider';

const fileName = '.changelog.yaml';
const load = (p: string): Options => yaml.safeLoad(fs.readFileSync(p, 'utf8')) || {};
const defaultConfig: Options = load(path.join(__dirname, '../../', fileName));

export type OptionValue = string | string[] | boolean | undefined;

export interface Option {
    [key: string]: Option | OptionValue;
}

export interface Options extends Option {
    config?: string;
    types?: string[];
    plugins?: string[];
    provider?: ProviderName;
    [key: string]: Option | OptionValue;
}

export default class Config {
    public readonly plugins: ReadonlyArray<string> = defaultConfig.plugins || [];
    public readonly types: ReadonlyArray<string>;
    public readonly provider: ProviderName;
    public readonly options: Option;

    public constructor(options?: Options) {
        let config: Options = Object.assign({}, defaultConfig, options || {});
        let { config: filePath } = config;

        if (typeof filePath === 'string') {
            filePath = path.resolve(process.cwd(), filePath);

            if (typeof filePath === 'string' && filePath.length && fs.existsSync(filePath)) {
                config = Object.assign({}, config, load(filePath));
            }
        }

        this.provider = config.provider || ProviderName.None;
        this.types = config.types || [];
        this.plugins = this.plugins.concat(config.plugins || []);
        this.options = new Proxy(config as Option, {
            get(target, name, receiver): Option | undefined {
                return Reflect.get(target, name, receiver);
            }
        });
    }
}
