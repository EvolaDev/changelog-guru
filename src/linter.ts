import { Task } from 'tasktree-cli/lib/task';
import { Commit } from './entities/commit';
import { PluginOption } from './config/config';
import { PluginLoader } from './plugins/plugin-loader';
import { State } from './state/state';
import { CommitPlugin } from './plugins/commit-plugin';
import Key from './utils/key';

export interface LinterOptions {
    lowercaseTypesOnly?: boolean;
    maxHeaderLength?: number;
}

export interface LintOptions {
    header: string;
    body: string[];
    type: string;
    scope: string;
    subject: string;
}

export default class Linter {
    public static DEFAULT_HEADER_MAX_LENGTH = 100;
    public static MIN_SUBJECT_LENGTH = 6;
    public static EMPTY_VALUE = '';

    private task: Task;
    private plugins: [string, PluginOption][];
    private types: string[];
    private maxHeaderLength: number;
    private lowercaseTypesOnly: boolean;
    private state: State = new State();
    private pluginLoader: PluginLoader = new PluginLoader();

    public constructor(plugins: [string, PluginOption][], types: string[], task: Task, options: LinterOptions) {
        this.plugins = plugins;
        this.types = types;
        this.task = task;
        this.maxHeaderLength = options.maxHeaderLength || Linter.DEFAULT_HEADER_MAX_LENGTH;
        this.lowercaseTypesOnly = !!options.lowercaseTypesOnly;
    }

    public async lint(message: string = Linter.EMPTY_VALUE): Promise<void> {
        const [header, body] = this.parseMessage(message);
        const [type, scope, subject] = this.parseHeader(header);

        await Promise.all(
            this.plugins.map(
                ([name, config]): Promise<void> =>
                    this.lintWithPlugin(name, config, {
                        header,
                        body,
                        type,
                        scope,
                        subject,
                    })
            )
        );
    }

    private parseMessage(message: string): [string, string[]] {
        let lines: string[] = [];
        let header = Linter.EMPTY_VALUE;

        if (message) {
            lines = message.split(Commit.LINE_SEPARATOR);
            header = lines.shift() as string;
        } else {
            this.task.error('Empty commit message');
        }

        return [header, lines];
    }

    private parseHeader(header: string): [string, string, string] {
        const [type, scope, subject] = Commit.splitHeader(header);
        const { task, maxHeaderLength } = this;
        const clearType = Key.unify(type);

        if (header.length > maxHeaderLength) task.error(`Header is longer than {bold ${maxHeaderLength}} characters`);

        if (type) {
            if (this.lowercaseTypesOnly && type !== clearType) task.error('Type is not in lowercase');
            if (!this.types.some((name): boolean => name === clearType)) task.error('Unknown commit type!');
        } else {
            task.error('Type is not defined or is not separated from the subject with "{bold :}"');
        }

        if (!subject) task.error('Subject is empty');
        if (subject.length < Linter.MIN_SUBJECT_LENGTH) task.error('Subject is not informative');

        return [clearType || Linter.EMPTY_VALUE, scope || Linter.EMPTY_VALUE, subject || Linter.EMPTY_VALUE];
    }

    private async lintWithPlugin(name: string, config: PluginOption, options: LintOptions): Promise<void> {
        const { task } = this;
        const plugin = await this.pluginLoader.load(task, { name, config, context: this.state });

        if (plugin instanceof CommitPlugin) {
            (plugin as CommitPlugin).lint(options, task);
        }
    }
}
