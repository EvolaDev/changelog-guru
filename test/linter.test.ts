import stripAnsi from 'strip-ansi';
import { Theme } from 'tasktree-cli/lib/theme';
import { Task } from 'tasktree-cli/lib/task';
import { ConfigLoader } from '../src/config/config-loader';
import { Config } from '../src/config/config';
import { MockLinter } from './__mocks__/linter.mock';

describe('Linter', (): void => {
    let $task: Task;
    let $config: Config;

    beforeAll((done): void => {
        const loader = new ConfigLoader();

        loader.load().then((config): void => {
            $config = config;

            expect(config).toBeDefined();
            done();
        });
    });

    beforeEach((): void => {
        $task = new Task('Lint commit header');
    });

    it('Default', (done): void => {
        const linter = new MockLinter($task, {
            config: { lowercaseTypesOnly: true, maxHeaderLength: MockLinter.DEFAULT_HEADER_MAX_LENGTH },
            plugins: $config.getPlugins(),
            types: $config.getTypes().map(([name]): string => name),
        });

        Promise.all([
            linter.lint('test(core): some subject message 1'),
            linter.lint('test: some subject message 2'),
        ]).then((): void => {
            expect($task.haveErrors()).toBeFalsy();

            done();
        });
    });

    it('Incorrect commit messages', (done): void => {
        const linter = new MockLinter($task, {
            config: { lowercaseTypesOnly: true, maxHeaderLength: MockLinter.DEFAULT_HEADER_MAX_LENGTH },
            plugins: $config.getPlugins(),
            types: $config.getTypes().map(([name]): string => name),
        });

        Promise.all([linter.lint(''), linter.lint('Test:'), linter.lint('wow: some subject message')]).then(
            (): void => {
                const theme = new Theme();
                const snapshot = $task.render(theme).map((line): string => stripAnsi(line));

                expect($task.haveErrors()).toBeTruthy();
                expect(snapshot).toMatchSnapshot();

                done();
            }
        );
    });
});
