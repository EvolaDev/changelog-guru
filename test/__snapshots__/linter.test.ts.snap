// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Linter Incorrect commit messages 1`] = `
Array [
  "  ⠙ Lint commit header",
  "  ─ ✖ Empty commit message",
  "  ─ ✖ Type is not defined or is not separated from the subject with \\":\\"",
  "  ─ ✖ Subject is empty",
  "  ─ ✖ Subject is not informative",
  "  ─ ✖ Type is not defined or is not separated from the subject with \\":\\"",
  "  ─ ✖ Subject is empty",
  "  ─ ✖ Subject is not informative",
  "  ─ ✖ Unknown commit type!",
  "  ─ ✖ Commit type  is not assigned with section",
  "  ─ ✖ Commit type  is not assigned with section",
  "  ─ ✖ Commit type wow is not assigned with section",
  "  ─ ℹ Header: undefined",
  "  ─ ℹ Header: Test:",
  "  ─ ℹ Header: wow: some subject message",
]
`;
